src/fsharp/FSharp.Core/control.fs

```diff
@@ -927,7 +950,11 @@ namespace Microsoft.FSharp.Control
                         token
                         trampolineHolder
                         (fun res -> resultCell.RegisterResult(Ok(res),reuseThread=true))
-                        (fun edi -> resultCell.RegisterResult(Error(edi),reuseThread=true))
+                        (fun edi ->
+                            let result =
+                                if token.IsCancellationRequested then Canceled(protectExn (new OperationCanceledException()) edi)
```

Comment:
Likewise here?

Reply:
see above. We have to do this for every relevant element